src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java

```diff
@@ -0,0 +1,223 @@
+import net.minecraft.util.hit.BlockHitResult;
+import org.joml.Vector2i;
+import org.joml.Vector2ic;
+
+import java.util.Iterator;
+import java.util.List;
+
+public class ActionViewer {
+    private static Action action = null;
+    private static ReferenceBook book = null;
+
+    private static int scroll = 0;
+    private static boolean tall = false;
+    private static boolean hover = false;
+
+    public static void invalidate() {
+        action = null;
+        book = null;
+        scroll = 0;
+        tall = false;
+        hover = false;
+    }
+    public static boolean isValid() {
+        return action != null || book != null;
+    }
+
+    public static void onClickChest(BlockHitResult hitResult) {
+        invalidate();
+        var world = CodeClient.MC.world;
+        if (world == null || !Config.getConfig().ActionViewer) return;
+        var position = hitResult.getBlockPos();
+        if (CodeClient.location instanceof Dev dev && dev.isInDev(position) && world.getBlockEntity(position) instanceof ChestBlockEntity) {
+            var signEntity = world.getBlockEntity(position.down().west());
+            if (signEntity instanceof SignBlockEntity sign) {
+                try {
+                    action = ActionDump.getActionDump().findAction(sign.getFrontText());
+                } catch (Exception ignored) {
+
+                }
+            }
+        }
+    }
+
+    public static List<Text> getOverlayText() {
+        if (CodeClient.location instanceof Dev dev) {
+            if (!Config.getConfig().ActionViewer) return null;
+            if (book != null) return book.getTooltip();
+            if (action == null) {
+                var book = dev.getReferenceBook();
+                if (book == null || book.isEmpty()) return null;
+                ActionViewer.book = book;
+                return book.getTooltip();
+            }
+            var item = action.icon.getItem();
+            return item.getTooltip(null, TooltipContext.BASIC);
+        }
+        return null;
+    }
+
+    public static void render(DrawContext context, int mouseX, int mouseY, HandledScreen<?> handledScreen) {
+        if (!(handledScreen instanceof GenericContainerScreen || handledScreen instanceof CustomChestMenu))
+            return;
+        var text = getOverlayText();
+        if (text == null) return;
+
+        ActionTooltipPositioner.INSTANCE.setMousePosition(mouseX,mouseY); // I would pass these through like normal, but draw tooltip might be moved to a utility class in the future.
+
+        var z = hover ? 500 : 300;
+        drawTooltip(context, handledScreen, transformText(text), 176, scroll, z);
+    }
+
+    public static boolean scroll(Screen screen, double mouseX, double mouseY, double horizontalAmount, double verticalAmount) {
+        if (!(screen instanceof GenericContainerScreen || screen instanceof CustomChestMenu)) return false;
+        if (isValid() && tall && hover) {
+            scroll -= 8 * (int) verticalAmount;
```

Comment:
scroll is inverted and can't scroll to the very top or bottom

Reply:
ok scrolling to the top or bottom is a problem with the large custom chest on big enough GUI scales, since that menu goes off the screen. We could ignore this, or account for it.

Reply:
changed scroll inversion to be a config option (as for me it felt more natural) and fixed the scrolling issue on the large custom chest.